nemo_aligner/utils/trt_llm.py

```diff
@@ -55,32 +70,44 @@ def __init__(
         self.unload_engine_train = unload_engine_train
         self.trt_model_type = trt_model_type
         self.reshard_model = reshard_model
-        self.pad_id = tokenizer.eos_id
-
         self.trt_llm_exporter = TensorRTLLM(trt_model_dir, load_model=False)
         self._trtllm_model_compiled = False
 
         rng_generator = torch.Generator(device="cpu")
         seed = secrets.randbits(32) if seed is None else seed
         rng_generator.manual_seed(seed)
-
         self.rng_generator = rng_generator
 
+        self.pad_id = tokenizer.pad_id
+        end_id = tokenizer.eos_id
         end_strings = list(end_strings)
-        assert all(
-            "," not in end_string for end_string in end_strings
-        ), "detected `,` in a specified end_string. This is will cause an error when converting it into ids for TRT-LLM"
 
-        end_strings = [[",".join(end_strings)] for _ in range(self.generation_batch_size)]
+        # TRT-LLM uses different logic to compute the response length depending on if we specify stop_list or end_id
+        # we want to use the same logic(which is to include the stop token in response length) so we put end_id into the stop_list
+        # manually
+        if len(end_strings) == 0:
+            ids, offsets = [end_id], [1]
+        else:
+            assert all(
+                "," not in end_string for end_string in end_strings
+            ), "detected `,` in a specified end_string. This is will cause an error when converting it into ids for TRT-LLM"
+
+            end_strings = [[",".join(end_strings)] for _ in range(self.generation_batch_size)]
+            # use an arbitary ref string to obtain end_string ids
+            stop_list = tensorrt_llm_run.to_word_list_format(
+                end_strings, build_tokenizer(self.tokenizer), ref_str="green tea icecream"
+            )
+            # add the end_id, we only need to manipulate one slice
+            ids, offsets = stop_list[0].tolist()
```

Comment:
Would you happen to have some examples of `ids` / `offsets` to show? I don't remember what the format is like. I also can't figure it out easily with a Google search...

Reply:
```
array([[[251636, 251694,    426,   1821,   2382, 251694, 251621, 255001],
        [     7,      8,     -1,     -1,     -1,     -1,     -1,     -1]]],
      dtype=int32)
```
here's a sample stop_list

the first row is the stop token ids, and the second is the cumulative sum

in the above example
[251636, 251694,    426,   1821,   2382, 251694, 251621] is one sequence of stop tokens
and 255001 is the other one

the offset is 7,8 -> this means the first stop word is [0, 7) second is [7, 8)

Reply:
Ok thanks, can we `assert max(offsets) == len(ids)`? (I just want to be sure TRT-LLM doesn't change something later that would break our logic)

Reply:
e888d4f4a7e8bf31d6066d547af51ef19f85047a